File: cardviewplus/src/main/java/com/alibardide5124/cardviewplus/CardViewPlus.java
package com.alibardide5124.cardviewplus;

import android.animation.FloatEvaluator;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.MotionEvent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;

public class CardViewPlus extends CardView {


    private Context context;
    private float normalElevation = 5;
    private float pressedElevation = 10;
    private boolean isAnimationEnabled = true;
    private int clickDelay = 150;

    private int animationDuration = 150;
    private boolean isPressed = false;


    public CardViewPlus(@NonNull Context context) {
        super(context);
        init(context);
    }
    public CardViewPlus(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }
    public CardViewPlus(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(@NonNull Context context) {
        this.context = context;

        initElevation();
    }
    private void init(@NonNull Context context, @Nullable AttributeSet attrs) {
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.CardViewPlus, 0, 0);

        this.context = context;
        normalElevation = (typedArray.getDimension(R.styleable.CardViewPlus_normalElevation, normalElevation));
        pressedElevation = typedArray.getDimension(R.styleable.CardViewPlus_pressedElevation, normalElevation * 2);
        isAnimationEnabled = typedArray.getBoolean(R.styleable.CardViewPlus_animationEnabled, isAnimationEnabled);
        clickDelay = typedArray.getInt(R.styleable.CardViewPlus_clickMode, clickDelay);

        typedArray.recycle();
        initElevation();
    }
    private void initElevation() {
        if (pressedElevation > normalElevation)
            setMaxCardElevation(pressedElevation);
        else
            setMaxCardElevation(normalElevation);

        setCardElevation(normalElevation);
    }

    public boolean isAnimationEnabled() { return isAnimationEnabled; }
    public void setAnimationEnabled(boolean animationEnabled) { isAnimationEnabled = animationEnabled; }

    public float getNormalElevation() { return normalElevation; }
    public void setNormalElevation(float normalElevation) { this.normalElevation = normalElevation; }

    public float getPressedElevation() { return pressedElevation; }
    public void setPressedElevation(float pressedElevation) { this.pressedElevation = pressedElevation; }

    public int getClickDelay() { return clickDelay; }
    public void setClickDelay(int clickDelay) { this.clickDelay = clickDelay; }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        onTouchEvent(ev);
        return false;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (isAnimationEnabled) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    if (!isPressed) {
                        onTouchAnimation();
                        isPressed = true;
                    } break;
                case MotionEvent.ACTION_UP:
                    if (isPressed) {
                        onReleaseAnimation();
                        isPressed = false;
                        setClickable(false);
                        new Handler().postDelayed(new Runnable() {
                            @Override
                            public void run() { performClick(); }
                        }, clickDelay);
                        return false;
                    } break;
                case MotionEvent.ACTION_MOVE:
                    if (isPressed) {
                        float eventY = event.getY();
                        float eventX = event.getX();
                        if (eventY > getHeight() || eventX > getWidth() || eventY < 0 || eventX < 0) {
                            onReleaseAnimation();
                            isPressed = false;
                        } return false;
                    }
                    break;
                case MotionEvent.ACTION_CANCEL:
                    if (isPressed) {
                        onReleaseAnimation();
                        isPressed = false;
                    }
            }
        } return true;
    }

    private void onTouchAnimation() {
        ObjectAnimator.ofObject(
                this,
                "cardElevation",
                new FloatEvaluator(),
                normalElevation,
                pressedElevation
        ).setDuration(animationDuration)
                .start();
    }
    private void onReleaseAnimation() {
        ObjectAnimator.ofObject(
            this,
                "cardElevation",
                new FloatEvaluator(),
                pressedElevation,
                normalElevation
        ).setDuration(animationDuration)
                .start();
    }
}

File: cardviewplus/src/main/java/com/alibardide5124/cardviewplus/CardViewPlus.kt
package com.alibardide5124.cardviewplus

import android.animation.ArgbEvaluator
import android.animation.FloatEvaluator
import android.animation.ObjectAnimator
import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.cardview.widget.CardView

class CardViewPlus : CardView {

    var normalCardElevation = 5f
    var pressedCardElevation = 10f
    var cardPressedColor = cardBackgroundColor.defaultColor
    var isAnimationEnabled = true
    var clickDelay = 150
    private var isCardPressed = false
    private val animationDuration = 150
    private val defaultColor = cardBackgroundColor.defaultColor

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs)
    }
    
    private fun init(context: Context, attrs: AttributeSet? = null) {
        if (attrs != null) {
            val typedArray = context.obtainStyledAttributes(attrs, R.styleable.CardViewPlus, 0, 0)
            normalCardElevation = typedArray.getDimension(R.styleable.CardViewPlus_normalCardElevation, normalCardElevation)
            pressedCardElevation = typedArray.getDimension(R.styleable.CardViewPlus_pressedCardElevation, normalCardElevation * 2)
            cardPressedColor = typedArray.getColor(R.styleable.CardViewPlus_cardPressedColor, cardPressedColor)
            isAnimationEnabled = typedArray.getBoolean(R.styleable.CardViewPlus_animationEnabled, isAnimationEnabled)
            clickDelay = typedArray.getInt(R.styleable.CardViewPlus_clickDelay, clickDelay)
            typedArray.recycle()
        } 
        initElevation()
    }

    private fun initElevation() {
        maxCardElevation = if (pressedCardElevation > normalCardElevation) pressedCardElevation else normalCardElevation
        cardElevation = normalCardElevation
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        onTouchEvent(ev)
        return false
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (isAnimationEnabled) {
            when (event.action) {
                MotionEvent.ACTION_DOWN -> if (!isCardPressed) {
                    onTouchAnimation()
                    isCardPressed = true
                }
                MotionEvent.ACTION_UP -> if (isCardPressed) {
                    onReleaseAnimation()
                    isCardPressed = false
                    isClickable = false
                    Handler().postDelayed({ performClick() }, clickDelay.toLong())
                    return false
                }
                MotionEvent.ACTION_MOVE -> if (isCardPressed) {
                    val checkX = event.x < 0 || event.x > width
                    val checkY = event.y < 0 || event.y > height
                    if (checkX || checkY) {
                        onReleaseAnimation()
                        isCardPressed = false
                    }
                    return false
                }
                MotionEvent.ACTION_CANCEL -> if (isCardPressed) {
                    onReleaseAnimation()
                    isCardPressed = false
                }
            }
        } else
            super.onTouchEvent(event)
        return true
    }

    private fun onTouchAnimation() {
        touchAnimation().start()
        colorAnimation().start()
    }

    private fun onReleaseAnimation() {
        touchAnimation().reverse()
        colorAnimation().reverse()
    }

    private fun touchAnimation() : ObjectAnimator = ObjectAnimator.ofObject(
            this,
            "cardElevation",
            FloatEvaluator(),
            normalCardElevation,
            pressedCardElevation
    ).setDuration(animationDuration.toLong())

    private fun colorAnimation() : ObjectAnimator = ObjectAnimator.ofObject(
            this,
            "cardBackgroundColor",
            ArgbEvaluator(),
            defaultColor,
            cardPressedColor
    ).setDuration(animationDuration.toLong())
}